### PR TITLE
[cairo,cairomm] add fontconfig and quartz feature and update cairo version

### PR DESCRIPTION
**Pull request by abique**

### CODE REVIEW

**Review by JackBoosY — CHANGES_REQUESTED** [review]
Please bump the port version. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#update-the-version-field-in-the-control-file-of-any-modified-ports).
Also please run command `vcpkg x-add-version --all` then commit the changes.

**Comment by abique** [comment]
@JackBoosY Thank you very much for your feedback. I've updated the pull request accordingly.

**Comment by JackBoosY** [comment]
Please get failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=50919&view=artifacts&pathAsName=false&type=publishedArtifacts).

**Comment by abique** [comment]
To me it looks line a problem in the other packages.

**Comment by abique** [comment]
I've made fontconfig a feature, that way it can be disabled.

**Comment by abique** [comment]
> Please bump the port version. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#update-the-version-field-in-the-control-file-of-any-modified-ports).
> Also please run command `vcpkg x-add-version --all` then commit the changes.

Thank you for that, this is now addressed.

**Comment by JackBoosY** [comment]
Please get failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=51088&view=artifacts&pathAsName=false&type=publishedArtifacts).
The osgearth regression is not related to this PR.

Waiting for #16904 merged.

**Comment by abique** [comment]
I had to rework cairomm because it was trying to compile quartz on Linux and Windows which obviously will fail.

**Comment by JackBoosY** [comment]
Get failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=51195&view=artifacts&pathAsName=false&type=publishedArtifacts).

File: ports/cairo/vcpkg.json [hunk 1]
```diff
@@ -13,10 +14,15 @@
     "freetype"
   ],
   "features": {
+    "fontconfig": {
+      "description": "build with fontconfig",
+      "dependencies": [
+        "fontconfig"
+      ]
+    },
```
Comment (Hoikas) [inline ports/cairo/vcpkg.json:22]: This should be added to the default-features, otherwise users will find their feature set changing unexpectedly.

Comment (abique) [reply ports/cairo/vcpkg.json:22]: We want fontconfig only on Linux or if x11 is enabled right?
There is no point in enabling fontconfig on windows or osx

Comment (JackBoosY) [reply ports/cairo/vcpkg.json:22]: Therefore, it should be:
```json
        {
          "name": "fontconfig",
          "platform": "linux"
        }
```

Comment (abique) [reply ports/cairo/vcpkg.json:22]: Thank you, I've done that.

Comment (abique) [reply ports/cairo/vcpkg.json:26]: Actually this is illegal to put such a thing in default features:
```./vcpkg format-manifest ports/cairo/vcpkg.json 
Failed to parse manifest file: /home/abique/develop/vcpkg/ports/cairo/vcpkg.json
Errors occurred while parsing /home/abique/develop/vcpkg/ports/cairo/vcpkg.json
    $.default-features[1]: mismatched type: expected an identifier
```

**Comment by JackBoosY** [comment]
Please use command `./vcpkg format-manifest --all` then commit changes.

File: ports/cairo/vcpkg.json (hunk 1 above)
Comment (abique) [reply ports/cairo/vcpkg.json:23]: I've made fontconfig a default feature as you requested.

File: ports/cairomm/vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,78 @@
+{
+  "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json",
+  "name": "cairomm",
+  "version-string": "1.16.0",
+  "port-version": 6,
```
Comment (JackBoosY) [inline ports/cairomm/vcpkg.json:5]:
```suggestion
```

Since this port is updating, please remove this.

File: ports/cairo/vcpkg.json [hunk 3]
```diff
@@ -10,13 +11,19 @@
...
   ],
   "default-features": [
+    "fontconfig",
     "freetype"
   ],
   "features": {
+    "fontconfig": {
+      "description": "build with fontconfig",
+      "dependencies": [
+        "fontconfig"
```
Comment (JackBoosY) [inline ports/cairo/vcpkg.json:21]:
```suggestion
        {
          "name": "fontconfig",
          "platform": "linux"
        }
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by abique** [comment]
LGTM!

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
I hope my colleague will do further review.

**Comment by Neumann-A** [comment]
this PR collides with #13100. Also could have switched the build to the native buildsystem via `vcpkg_configure_make` which will be required anyway.

**Comment by abique** [comment]
> this PR collides with #13100. Also could have switched the build to the native buildsystem via `vcpkg_configure_make` which will be required anyway.
#13100 is massive, and it does not address the same issues. While this one has a minimum amount of changes and brings cairo into a better shape. You now can have freetype without using fontconfig, and it provide better support for quartz on mac.

I'd merge #16825 first.

**Comment by Neumann-A** [comment]
Yeah but the question is if the native buildsystem supports the new features introduced here. If not this PR has to remove the features added here.

**Comment by abique** [comment]
Of course you can disable fontconfig when using the configure script.

**Comment by abique** [comment]
I've bumped cairo to 17.4 because 1.16.0 had a bunch of known bugs.

**Comment by Hoikas** [comment]
A previous cairo update PR was shelved because 1.17.x are snapshots. https://github.com/microsoft/vcpkg/pull/16036#discussion_r570077146

**Comment by abique** [comment]
I have a crash in cairo, which I'll confirm soon if it is fixed by 1.17.4.
I believe 1.16.0 to be less stable than 1.17.4 despite the snapshot denomination.

**Comment by abique** [comment]
To be precise we're running into https://gitlab.freedesktop.org/cairo/cairo/-/issues/412

**Comment by JackBoosY** [comment]
Should we apply [this patch](https://gitlab.freedesktop.org/cairo/cairo/-/merge_requests/44/diffs)?

**Comment by abique** [comment]
After some testing we figured that even 1.17.4 did not fix our issues on macos.
We are going to check if switching from static linking to dynamic linking fixes it.
Sorry for the noise, I'll come back with the conclusion.
Through, archlinux is using 1.17.4 and if that version was broken we'd know it. I've checked the changelog and basically the big change is the new backend cogl. Appart from that it is mostly fixes.

**Comment by JackBoosY** [comment]
> After some testing we figured that even 1.17.4 did not fix our issues on macos.
> We are going to check if switching from static linking to dynamic linking fixes it.
> Sorry for the noise, I'll come back with the conclusion.
> Through, archlinux is using 1.17.4 and if that version was broken we'd know it. I've checked the changelog and basically the big change is the new backend cogl. Appart from that it is mostly fixes.

So should we wait for upstream fix this issue completely?

**Comment by abique** [comment]
> 
> 
> > After some testing we figured that even 1.17.4 did not fix our issues on macos.
> > We are going to check if switching from static linking to dynamic linking fixes it.
> > Sorry for the noise, I'll come back with the conclusion.
> > Through, archlinux is using 1.17.4 and if that version was broken we'd know it. I've checked the changelog and basically the big change is the new backend cogl. Appart from that it is mostly fixes.
> 
> So should we wait for upstream fix this issue completely?

We don't understand why, but switching from static build to dynamic build seems to fix it.
We'll get back with more info.

**Comment by abique** [comment]
We've been using 1.17.4 for some time now on Windows and Macos and it works like a charm. Plus Archlinux is using this version since a lot of time.

I think it is safe, and the right move to jump to this version.

**Comment by abique** [comment]
As you don't feel excited about cairo-1.17.4 I'll revert to 1.16.0 so at least our work is not lost.

**Comment by abique** [comment]
@PhoebeHui Hi,
Shall we merge it?

**Comment by JackBoosY** [comment]
Ping @strega-nil for review and merge this PR.

**Comment by strega-nil** [comment]
Thanks @abique :)

**Pull request merged**